security/src/main/java/com/networknt/security/JwtVerifier.java

```diff
@@ -302,6 +304,67 @@ public JwtClaims verifyJwt(String jwt, boolean ignoreExpiry, boolean isToken, St
+        ClientConfig clientConfig = ClientConfig.get();
+        String configuredAudience = null;
+        try {
+            if (requestPath == null && jwkServiceIds == null) {
+                // if both requestPath and jwkServiceIds are null, it means the single oauth server is used.
+                configuredAudience = audience;
+                if (configuredAudience != null) {
+                   // validate the audience against the configured audience in the client.yml
+                   boolean r = isJwtAudienceValid(claims, configuredAudience);
+                   if(!r) {
+                       throw new InvalidJwtException("Invalid Audience", Collections.singletonList(new ErrorCodeValidator.Error(ErrorCodes.AUDIENCE_INVALID, "Invalid Audience")), context);
+                   }
+                }
+            } else if (requestPath != null) {
+                String serviceId = getServiceIdByRequestPath(clientConfig, requestPath);
+                if(serviceId == null && audienceMap != null && audienceMap.size() > 0) {
+                    configuredAudience = audienceMap.get(serviceId);
+                    boolean r = isJwtAudienceValid(claims, configuredAudience);
+                    if(!r) {
+                        throw new InvalidJwtException("Invalid Audience", Collections.singletonList(new ErrorCodeValidator.Error(ErrorCodes.AUDIENCE_INVALID, "Invalid Audience")), context);
+                    }
+                }
+            } else {
+                if(jwkServiceIds != null && jwkServiceIds.size() > 0) {
+                    // more than one serviceIds are passed in from the UnifiedSecurityHandler. Just use each serviceId to get the audience.
+                    for(String serviceId: jwkServiceIds) {
+                        if(audienceMap != null && audienceMap.size() > 0) {
+                            configuredAudience = audienceMap.get(serviceId);
+                            boolean r = isJwtAudienceValid(claims, configuredAudience);
+                            if(!r) {
+                                throw new InvalidJwtException("Invalid Audience", Collections.singletonList(new ErrorCodeValidator.Error(ErrorCodes.AUDIENCE_INVALID, "Invalid Audience")), context);
+                            }
+                        }
+                    }
+                }
+            }
+        } catch (MalformedClaimException e) {
+            logger.error("MalformedClaimException:", e);
+            throw new InvalidJwtException("MalformedClaimException", new ErrorCodeValidator.Error(ErrorCodes.MALFORMED_CLAIM, "Invalid Audience"), e, context);
+        }
+    }
+
+    private boolean isJwtAudienceValid(JwtClaims claims, String audience) throws MalformedClaimException {
+        if (claims.getAudience() == null) {
+            return false;
+        }
+        if (claims.getAudience().size() == 1) {
```

Comment:
The claims.getAudience() returns a list of strings. 

```
        if (claims.getAudience() == null) {
            return false;
        }
        if (claims.getAudience().size() == 1) {
            return claims.getAudience().get(0).equals(audience);
        }
        return claims.getAudience().contains(audience);

```
If the list is null, return false. 
If the size is 1, there is only one string in the list: the single audience case. Just compare it with the passed-in audience.
The last return is the multiple audiences case. It will return true if any of the audience matches the passed-in audience. 

I think the above code should cover all the scenarios. Let me know if I missed anything. Thanks.

Reply:
Sorry I think I got confused. So do we really need a special handling for when size is 1? `claims.getAudience().contains(audience)` should cover everything other than null, no?

Reply:
It is optional. It is just performance improvement as, in most cases, we only have one audience. The list contains() is slower than get(0).